hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/FSEditLog.java

```diff
@@ -1666,6 +1670,9 @@ synchronized void recoverUnclosedStreams() {
     } catch (IOException ex) {
       // All journals have failed, it is handled in logSync.
       // TODO: are we sure this is OK?
+      if (terminateOnFailure) {
+        throw ex;
```

Comment:
In my [last comment about this](https://github.com/apache/hadoop/pull/4744#discussion_r1026931795) I was thinking that we should have code more similar to `FSEditLog#startLogSegment()` where we actually call `terminate()` as opposed to just throwing the exception. I was thinking that we would always at least throw the exception, but only _terminate_ when `terminateOnFailure` is true, like:
```java
} catch (IOException ex) {
  if (terminateOnFailure) {
    final String msg = "Unable to recover log segments: too few journals successfully recovered.";
    LOG.error(msg, ex);
    synchronized (journalSetLock) { 
      IOUtils.cleanupWithLogger(LOG, journalSet);
    }
    terminate(1, msg);
  } else {
    throw ex;
  }
}
```

What do you think? I am open to other options but this is what I had in mind.

Reply:
@xkrogen Yeah. After deep think, this is a better solution. Thanks so much, sir. I learned a lot from you.

I have updated the code, please help me review it again. Thanks.